src/auto-evo/mutation_strategy/UpgradeOrganelle.cs

```diff
@@ -60,31 +83,60 @@ public UpgradeOrganelle(Func<OrganelleDefinition, bool> criteria, string upgrade
         {
             var mutated = new List<Tuple<MicrobeSpecies, double>>();
 
-            var newSpecies = (MicrobeSpecies)baseSpecies.Clone();
+            var eligibleOrganelles = new List<OrganelleTemplate>();
 
-            organelleList = newSpecies.Organelles.Organelles;
-            count = organelleList.Count;
-            for (var i = 0; i < count; ++i)
+            foreach (var organelle in baseSpecies.Organelles.Organelles)
             {
-                var organelle = organelleList[i];
-
                 if (allOrganelles.Contains(organelle.Definition))
+                    eligibleOrganelles.Add(organelle);
+            }
+
+            var eligibleOrganelleSample = eligibleOrganelles
+                .OrderBy(_ => random.Next())
+                .Take(Constants.AUTO_EVO_ORGANELLE_UPGRADE_ATTEMPTS);
```

Comment:
Is this random sampling really necessary as this seems like the reason why a temporary list needs to be allocated here?